lib/src/utils/decodeText.ts

```diff
@@ -0,0 +1,153 @@
+import type { DecodeTextOptions } from './DecodeTextOptions.js';
+import { UTF_16 } from './UTF_16.js';
+import { UTF_16_BE } from './UTF_16_BE.js';
+import { UTF_16_LE } from './UTF_16_LE.js';
+import { UTF_8 } from './UTF_8.js';
+
+/**
+ * Converts an ArrayBuffer or ArrayBufferView to a string. Similar to `TextDecoder.decode`
+ * but with a fallback for environments that don't support `TextDecoder`.
+ *
+ * @param data - The data to decode.
+ * @param options - The options for the decoding.
+ * @returns The string representation of the ArrayBuffer.
+ *
+ * @group Utils
+ *
+ * @beta
+ *
+ * @example
+ * {@includeCode ../../test/utils/decodeText.test.ts#example}
+ */
+export function decodeText(data: ArrayBuffer | ArrayBufferView<ArrayBuffer>, options: DecodeTextOptions = {}): string {
+	let view: DataView<ArrayBuffer>;
+
+	if (data instanceof ArrayBuffer) {
+		view = new DataView(data);
+	}
+	else {
+		view = new DataView(data.buffer, data.byteOffset, data.byteLength);
+	}
+
+	let byteOffset = 0;
+	let { encoding } = options;
+
+	// If no encoding is provided, try to detect it from the BOM
+	if (!encoding) {
+		const first = view.getUint8(0);
+		const second = view.getUint8(1);
+
+		// UTF-8 BOM
+		if (first == 0xef && second == 0xbb && view.getUint8(2) == 0xbf) {
+			encoding = UTF_8;
+			byteOffset = 3;
+		}
+		// UTF-16 BE BOM
+		else if (first == 0xfe && second == 0xff) {
+			encoding = UTF_16_BE;
+			byteOffset = 2;
+		}
+		// UTF-16 LE BOM
+		else if (first == 0xff && second == 0xfe) {
+			encoding = UTF_16_LE;
+			byteOffset = 2;
+		}
+		else {
+			encoding = UTF_8;
+		}
+	}
+
+	if (typeof TextDecoder !== 'undefined') {
+		return new TextDecoder(encoding).decode(view);
+	}
+
+	const { byteLength } = view;
+	const endian = encoding !== UTF_16_BE;
+	let str = '';
+	let char: number;
+
+	while (byteOffset < byteLength) {
+		switch (encoding) {
+			case UTF_8:
+				char = view.getUint8(byteOffset);
+
+				// Single byte (ASCII)
+				if (char < 128) {
+					byteOffset++;
+				}
+				// 2-byte sequence
+				else if (char >= 194 && char <= 223) {
+					if (byteOffset + 1 < byteLength) {
+						const byte2 = view.getUint8(byteOffset + 1);
+						if (byte2 >= 128 && byte2 <= 191) {
```

Comment:
This condition repeats 5-6 times later, is it worth creating utility function for it?

Reply:
It's slightly faster inline, so it would be a trade off.

Reply:
Right, makes sense